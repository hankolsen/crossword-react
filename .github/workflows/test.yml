name: melodikryzzet-tests

on: 
  push:
    branches:
      - 'master'
  pull_request:


jobs:
  test:
    name: 'Node ${{ matrix.node }}'
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node: [12, 14]
    env:
      SKIP_PREFLIGHT_CHECK: true
    steps:
    - uses: actions/checkout@v2
    - uses: actions/setup-node@v1
      with: 
        node-version: ${{ matrix.node }}
    - name: Install dependencies
      uses: bahmutov/npm-install@v1
      with: 
        install-command: yarn
    - name: Run tests
      run: yarn test:ci