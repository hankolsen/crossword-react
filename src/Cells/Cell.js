import React from 'react';
import PropTypes from 'prop-types';

import CellRectangle from './CellRectangle/CellRectangle';
import CellText from './CellText/CellText';
import ArrowAcross from './ArrowAcross';
import ArrowDown from './ArrowDown';
import CellLabel from './CellLabel/CellLabel';

const Cell = ({
  row,
  number = '',
  column,
  letter = '',
  arrow = '',
  clickHandler,
  highlighted,
  selected,
}) => (
  <g onClick={() => clickHandler(row, column)}>
    <CellRectangle
      column={column}
      row={row}
      selected={selected}
      highlighted={highlighted}
    />
    {number ? <CellLabel column={column} row={row} number={number} /> : null}
    {arrow && arrow === 'across' ? (
      <ArrowAcross column={column} row={row} />
    ) : null}
    {arrow && arrow === 'down' ? <ArrowDown column={column} row={row} /> : null}
    <CellText row={row} column={column} text={letter} />
  </g>
);

Cell.defaultProps = {
  number: '',
  letter: '',
  arrow: '',
};

Cell.defaultProps = {
  highlighted: false,
  selected: false,
};

Cell.propTypes = {
  row: PropTypes.number.isRequired,
  column: PropTypes.number.isRequired,
  number: PropTypes.number,
  letter: PropTypes.string,
  arrow: PropTypes.string,
  clickHandler: PropTypes.func.isRequired,
  highlighted: PropTypes.bool,
  selected: PropTypes.bool,
};

export default Cell;
