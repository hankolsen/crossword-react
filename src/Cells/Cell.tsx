import React from 'react';

import CellRectangle from './CellRectangle/CellRectangle';
import CellText from './CellText/CellText';
import ArrowAcross from './ArrowAcross';
import ArrowDown from './ArrowDown';
import CellLabel from './CellLabel/CellLabel';

type Props = {
  arrow?: string;
  clickHandler: (row: number, column: number) => void;
  column: number;
  highlighted?: boolean;
  letter?: string;
  number?: number;
  row: number;
  selected?: boolean;
};
const Cell = ({
  row,
  number,
  column,
  letter,
  arrow,
  clickHandler,
  highlighted,
  selected,
}: Props) => (
  <g onClick={() => clickHandler(row, column)} data-testid="cell">
    <CellRectangle
      column={column}
      row={row}
      selected={selected}
      highlighted={highlighted}
    />
    {number ? <CellLabel column={column} row={row} number={number} /> : null}
    {arrow && arrow === 'across' ? (
      <ArrowAcross column={column} row={row} />
    ) : null}
    {arrow && arrow === 'down' ? <ArrowDown column={column} row={row} /> : null}
    <CellText row={row} column={column} text={letter} />
  </g>
);

export default Cell;
