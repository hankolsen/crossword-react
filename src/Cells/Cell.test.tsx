import React from 'react';
import userEvent from '@testing-library/user-event';

import { renderWithSvg, screen } from '../test/test-utils';
import Cell from './Cell';

describe('Cell test', () => {
  it('should render an empty Cell', () => {
    const { container } = renderWithSvg(
      <Cell column={0} row={0} clickHandler={() => {}} />,
    );
    expect(
      container.querySelector('.crossword__cell-number'),
    ).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render a Cell with a letter, a number, and an across arrow', () => {
    const { container } = renderWithSvg(
      <Cell
        column={0}
        row={0}
        clickHandler={() => {}}
        letter="l"
        number={3}
        arrow="across"
      />,
    );
    expect(screen.queryByText(/^l$/i)).toBeInTheDocument();
    expect(screen.queryByText(/^q$/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/^3$/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render a Cell with a down arrow', () => {
    const { container } = renderWithSvg(
      <Cell column={0} row={0} clickHandler={() => {}} arrow="down" />,
    );
    expect(screen.queryByText(/^q$/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/^3$/i)).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should trigger click the click handler', () => {
    const mockClickHandler = jest.fn();
    const row = 1;
    const column = 2;

    renderWithSvg(
      <Cell
        column={column}
        row={row}
        clickHandler={mockClickHandler}
        arrow="down"
      />,
    );

    userEvent.click(screen.getByTestId('cell'));
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
    expect(mockClickHandler).toHaveBeenCalledWith(row, column);
  });
});
