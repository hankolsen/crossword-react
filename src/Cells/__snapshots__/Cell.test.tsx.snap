// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell test should render a Cell with a down arrow 1`] = `
<div>
  <svg>
    <g
      data-testid="cell"
    >
      <rect
        class="crossword__cell  "
        height="31"
        width="31"
        x="1"
        y="1"
      />
      <g>
        <line
          stroke="black"
          stroke-width="1"
          x1="29"
          x2="29"
          y1="4"
          y2="11"
        />
        <line
          stroke="black"
          stroke-width="1"
          x1="29"
          x2="22"
          y1="4"
          y2="4"
        />
        <path
          d="M27,11 L31,11 L29,16 z"
          fill="#000"
        />
      </g>
      <text
        class="crossword__cell-text"
        text-anchor="middle"
        x="15.5"
        y="25"
      />
    </g>
  </svg>
</div>
`;

exports[`Cell test should render a Cell with a letter, a number, and an across arrow 1`] = `
<div>
  <svg>
    <g
      data-testid="cell"
    >
      <rect
        class="crossword__cell  "
        height="31"
        width="31"
        x="1"
        y="1"
      />
      <text
        class="crossword__cell-number"
        x="2"
        y="10"
      >
        3
      </text>
      <g>
        <line
          stroke="black"
          stroke-width="1"
          x1="5"
          x2="5"
          y1="22"
          y2="29"
        />
        <line
          stroke="black"
          stroke-width="1"
          x1="5"
          x2="12"
          y1="29"
          y2="29"
        />
        <path
          d="M12,27 L12,31 L17,29 z"
          fill="#000"
        />
      </g>
      <text
        class="crossword__cell-text"
        text-anchor="middle"
        x="15.5"
        y="25"
      >
        l
      </text>
    </g>
  </svg>
</div>
`;

exports[`Cell test should render an empty Cell 1`] = `
<div>
  <svg>
    <g
      data-testid="cell"
    >
      <rect
        class="crossword__cell  "
        height="31"
        width="31"
        x="1"
        y="1"
      />
      <text
        class="crossword__cell-text"
        text-anchor="middle"
        x="15.5"
        y="25"
      />
    </g>
  </svg>
</div>
`;
